File: src/clase-04/practica/typescript/factory_starter.ts
// Actividad 2 — Factory Method (TypeScript)
// Objetivo: Implementar una UserFactory que cree distintos tipos de usuarios.
//
// Requisitos:
// - Interfaz User con método permissions(): string[]
// - Clases concretas: AdminUser, MemberUser, GuestUser
// - Factory: createUser(kind: string): User
// - Demostración: crear usuarios y mostrar permisos

interface User {
  permissions(): string[];
}

// TODO: implementar clases concretas
// class AdminUser implements User { ... }
// class MemberUser implements User { ... }
// class GuestUser implements User { ... }

class UserFactory {
  createUser(kind: string): User {
    const k = kind.toLowerCase();
    // TODO: devolver la clase correcta según k
    // if (k === "admin") return new AdminUser();
    // if (k === "member") return new MemberUser();
    // if (k === "guest") return new GuestUser();
    throw new Error("TODO: implementar createUser");
  }
}

// Uso (descomentar al completar)
// const f = new UserFactory();
// const u1 = f.createUser("admin");
// const u2 = f.createUser("member");
// const u3 = f.createUser("guest");
// console.log(u1.permissions(), u2.permissions(), u3.permissions());

File: src/clase-04/practica/python/factory_starter.py
"""
Actividad 2 — Factory Method (Python)
Objetivo: Implementar una UserFactory que cree distintos tipos de usuarios.

Requisitos:
- Interfaz/Clase base: User (método permissions() -> list[str])
- Clases concretas: AdminUser, MemberUser, GuestUser
- Factory: UserFactory.create_user(kind: str) -> User
- Demostración: crear usuarios y mostrar sus permisos.
"""
from abc import ABC, abstractmethod
from typing import List

class User(ABC):
    @abstractmethod
    def permissions(self) -> List[str]:
        pass

# TODO: implementar clases concretas con permisos apropiados
# class AdminUser(User): ...
# class MemberUser(User): ...
# class GuestUser(User): ...

class UserFactory:
    def create_user(self, kind: str) -> User:
        kind_lower = kind.lower()
        # TODO: devolver la clase correcta según kind_lower
        # if kind_lower == "admin": return AdminUser()
        # elif kind_lower == "member": return MemberUser()
        # elif kind_lower == "guest": return GuestUser()
        # else: raise ValueError(...)
        raise NotImplementedError("Completar la fábrica")

if __name__ == "__main__":
    # TODO: crear 'admin', 'member', 'guest' y mostrar sus permisos
    pass
